File: build_model/source/remake_packmol.py
#!/usr/bin/env python3

import re
import sys
import os
import subprocess
from typing import List


def read_mol_dat(mol_file: str) -> List[str]:
    """
    Read mol.dat file and extract molecule names (one per line).
    Ignore empty lines and comments starting with '#' or ';'.
    """
    if not os.path.exists(mol_file):
        raise FileNotFoundError(f"mol.dat file not found: {mol_file}")

    names = []
    with open(mol_file, 'r') as f:
        for line_num, line in enumerate(f, 1):
            stripped = line.strip()
            if not stripped or stripped.startswith('#') or stripped.startswith(';'):
                continue
            parts = stripped.split()
            if not parts:
                continue
            mol_name = parts[0]
            if not mol_name:
                print(f"Warning: Empty molecule name at line {line_num}")
                continue
            names.append(mol_name)
    return names

def modify_packmol(inp_file: str, output_file: str, target_mols: List[str], factor: float):
    """
    Modify the 'number' field in packmol.inp for specified molecules.
    Molecule name is extracted from the .pdb filename in 'structure' block.
    """
    if not os.path.exists(inp_file):
        raise FileNotFoundError(f"Input packmol file not found: {inp_file}")

    with open(inp_file, 'r') as f:
        lines = f.readlines()

    output_lines = []
    in_structure = False
    current_pdb = None

    for line in lines:
        stripped = line.strip()

        # Match start of structure block
        pdb_match = re.match(r'^\s*structure\s+([^#\s]+\.pdb)', stripped, re.IGNORECASE)
        if pdb_match:
            in_structure = True
            current_pdb = pdb_match.group(1)
            output_lines.append(line)
            continue

        # Match end of structure block
        if in_structure and stripped.startswith('end structure'):
            in_structure = False
            current_pdb = None
            output_lines.append(line)
            continue

        # Inside structure block, process 'number'
        if in_structure and stripped.startswith('number'):
            try:
                old_count = int(stripped.split()[1])
            except (IndexError, ValueError) as e:
                print(f"[packmol] Warning: Invalid number line, skipping: {line.strip()}")
                output_lines.append(line)
                continue

            # Extract molecule name from pdb filename (e.g., A.pdb -> A)
            mol_name = re.sub(r'\.pdb$', '', current_pdb, flags=re.IGNORECASE)

            if mol_name in target_mols:
                new_count = int(round(old_count * factor))
                if new_count < 0:
                    raise ValueError(f"Invalid count after scaling: {new_count} for {mol_name}")
                indent = line[:line.find('number')]
                new_line = f"{indent}number {new_count}\n"
                output_lines.append(new_line)
                print(f"[packmol] {mol_name}: {old_count} -> {new_count} (x{factor:.3f})")
            else:
                output_lines.append(line)
        else:
            output_lines.append(line)

    with open(output_file, 'w') as f:
        f.writelines(output_lines)
    print(f"Modified packmol input written to: {output_file}")


def modify_topology(top_file: str, output_file: str, target_mols: List[str], factor: float):
    """
    Modify the molecule counts in [ molecules ] section of the topology (.top) file.
    """
    if not os.path.exists(top_file):
        raise FileNotFoundError(f"Topology file not found: {top_file}")

    with open(top_file, 'r') as f:
        lines = f.readlines()

    output_lines = []
    in_molecules = False

    for line in lines:
        stripped = line.strip()

        if stripped.startswith('[ molecules ]'):
            in_molecules = True
            output_lines.append(line)
            continue

        # End of [ molecules ] section when another directive starts
        if in_molecules and stripped.startswith('[') and 'molecules' not in stripped.lower():
            in_molecules = False

        if in_molecules:
            if not stripped or stripped.startswith(';'):
                output_lines.append(line)
                continue

            parts = stripped.split()
            if len(parts) >= 2:
                mol_name = parts[0]
                try:
                    old_count = int(parts[1])
                except ValueError:
                    print(f"[top] Warning: Invalid count in line: {line.strip()}")
                    output_lines.append(line)
                    continue

                if mol_name in target_mols:
                    new_count = int(round(old_count * factor))
                    if new_count < 0:
                        raise ValueError(f"Invalid count after scaling: {new_count} for {mol_name}")
                    # Preserve formatting as much as possible
                    new_line = f"  {mol_name:8s} {new_count:>8d}\n"
                    output_lines.append(new_line)
                    print(f"[top] {mol_name}: {old_count} -> {new_count} (x{factor:.3f})")
                else:
                    output_lines.append(line)
            else:
                output_lines.append(line)
        else:
            output_lines.append(line)

    with open(output_file, 'w') as f:
        f.writelines(output_lines)
    print(f"Modified topology written to: {output_file}")


def main():
    script_name = os.path.basename(__file__)  # 只取文件名，如 "script.py"
    iter = os.environ['iter']
    # 或者用 full path: __file__

    print(f"""
===================================================================================================
 ---------------------------------------- iter="{iter}" -----------------------------------------
【INFO】Starting pre-equilibration and density analysis workflow
Script    : {script_name}
===================================================================================================""")

    inp_file = os.environ['INP_FILE']
    top_file = os.environ['TOP']
    mol_name = os.environ['MOL_name']
    target_molecules = mol_name.split()
    # Read scaling factor from environment variable

    try:
        factor = float(sys.argv[1])
        if factor < 0:
            print("Error: SCALE_FACTOR must be non-negative")
            exit(1)
    except (IndexError, ValueError):
        print("Error: Please provide a valid non-negative scaling factor as the first argument")
        exit(1)

    # Modify packmol input
    try:
        modify_packmol(inp_file, inp_file, target_molecules, factor)
    except Exception as e:
        print(f"Error: Failed to modify packmol.inp: {e}")
        exit(1)

    # Modify topology file
    top_file = f"{top_file}.top"
    try:
        modify_topology(top_file, top_file, target_molecules, factor)
    except Exception as e:
        print(f"Error: Failed to modify topology file: {e}")
        exit(1)

    print("✅ All files modified successfully!")


if __name__ == '__main__':
    main()

File: build_model/source/remake_packmol.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <regex>
#include <cstdlib>
#include <cmath>
#include <iomanip>

using namespace std;

/* ================= Utility ================= */

static inline string trim(const string& s) {
    size_t start = s.find_first_not_of(" \t\r\n");
    if (start == string::npos) return "";
    size_t end = s.find_last_not_of(" \t\r\n");
    return s.substr(start, end - start + 1);
}

static inline vector<string> split(const string& s) {
    vector<string> tokens;
    istringstream iss(s);
    string tok;
    while (iss >> tok) tokens.push_back(tok);
    return tokens;
}

static inline bool contains(const vector<string>& v, const string& key) {
    for (const auto& x : v)
        if (x == key) return true;
    return false;
}

/* ================= Packmol ================= */

void modify_packmol(const string& inp_file,
                    const vector<string>& target_mols,
                    double factor)
{
    struct MolCount {
        int origin;
        int new_count;
    };
    vector<MolCount> data_mol_nums;
    pair<int, float> origin_ratio;
    ifstream fin(inp_file);
    if (!fin)
        throw runtime_error("Input packmol file not found: " + inp_file);

    vector<string> lines;
    string line;
    while (getline(fin, line))
        lines.push_back(line + "\n");
    fin.close();

    vector<string> output;
    bool in_structure = false;
    string current_pdb;

    regex structure_re(R"(^\s*structure\s+([^#\s]+\.pdb))",
                       regex::icase);

    for (const auto& raw : lines) {
        string stripped = trim(raw);
        smatch match;

        /* structure xxx.pdb */
        if (regex_search(stripped, match, structure_re)) {
            in_structure = true;
            current_pdb = match[1];
            output.push_back(raw);
            continue;
        }

        if (in_structure && stripped.find("end structure") == 0) {
            in_structure = false;
            current_pdb.clear();
            output.push_back(raw);
            continue;
        }

        if (in_structure && stripped.find("number") == 0) {
            auto parts = split(stripped);
            if (parts.size() < 2) {
                cerr << "[packmol] Warning: invalid number line: "
                     << stripped << endl;
                output.push_back(raw);
                continue;
            }

            int old_count;
            try {
                old_count = stoi(parts[1]);
            } catch (...) {
                cerr << "[packmol] Warning: invalid number value: "
                     << stripped << endl;
                output.push_back(raw);
                continue;
            }

            string mol = regex_replace(current_pdb,
                                       regex(R"(\.pdb$)", regex::icase),
                                       "");

            if (contains(target_mols, mol)) {
                int new_count = int(round(old_count * factor));

                if (new_count < 0)
                    throw runtime_error("Negative molecule count for " + mol);

                size_t pos = raw.find("number");
                string indent = (pos != string::npos) ? raw.substr(0, pos) : "";

                ostringstream oss;
                oss << indent << "number " << new_count << "\n";
                output.push_back(oss.str());

                cout << "[packmol] " << mol << ": "
                     << old_count << " -> "
                     << new_count << " (x"
                     << fixed << setprecision(3) << factor << ")\n";
            } else {
                output.push_back(raw);
            }
        }
        else {
            output.push_back(raw);
        }
    }

    ofstream fout(inp_file);
    for (const auto& l : output)
        fout << l;
    fout.close();

    cout << "Modified packmol input written to: "
         << inp_file << endl;
}

/* ================= Topology ================= */

void modify_topology(const string& top_file,
                     const vector<string>& target_mols,
                     double factor)
{
    ifstream fin(top_file);
    if (!fin)
        throw runtime_error("Topology file not found: " + top_file);

    vector<string> lines;
    string line;
    while (getline(fin, line))
        lines.push_back(line + "\n");
    fin.close();

    vector<string> output;
    bool in_molecules = false;

    for (const auto& raw : lines) {
        string stripped = trim(raw);

        if (stripped == "[ molecules ]") {
            in_molecules = true;
            output.push_back(raw);
            continue;
        }

        if (in_molecules &&
            stripped.size() > 0 &&
            stripped[0] == '[' &&
            stripped.find("molecules") == string::npos) {
            in_molecules = false;
        }

        if (in_molecules) {
            if (stripped.empty() || stripped[0] == ';') {
                output.push_back(raw);
                continue;
            }

            auto parts = split(stripped);
            if (parts.size() < 2) {
                output.push_back(raw);
                continue;
            }

            string mol = parts[0];
            int old_count;

            try {
                old_count = stoi(parts[1]);
            } catch (...) {
                cerr << "[top] Warning: invalid count: "
                     << stripped << endl;
                output.push_back(raw);
                continue;
            }

            if (contains(target_mols, mol)) {
                int new_count = int(round(old_count * factor));
                if (new_count < 0)
                    throw runtime_error("Negative molecule count for " + mol);

                ostringstream oss;
                oss << "  " << left << setw(8) << mol
                    << right << setw(8) << new_count << "\n";
                output.push_back(oss.str());

                cout << "[top] " << mol << ": "
                     << old_count << " -> "
                     << new_count << " (x"
                     << fixed << setprecision(3) << factor << ")\n";
            } else {
                output.push_back(raw);
            }
        }
        else {
            output.push_back(raw);
        }
    }

    ofstream fout(top_file);
    for (const auto& l : output)
        fout << l;
    fout.close();

    cout << "Modified topology written to: "
         << top_file << endl;
}

/* ================= main ================= */

int main(int argc, char* argv[])
{
    if (argc < 3) {
        cerr << "Usage: " << argv[0]
             << " SCALE_FACTOR packmol_name\n";
        return 1;
    }

    double factor = atof(argv[1]);
    if (factor < 0.0) {
        cerr << "Error: SCALE_FACTOR must be non-negative\n";
        return 1;
    }

    const char* iter = getenv("iter");
    const char* top_env = getenv("TOP");
    const char* mol_env = getenv("MOL_name");

    if (!iter || !top_env || !mol_env) {
        cerr << "Error: environment variables iter / TOP / MOL_name must be set\n";
        return 1;
    }

    vector<string> target_mols = split(mol_env);

    cout << "\n"
         << "===================================================================================================\n"
         << " ---------------------------------------- iter=\"" << iter << "\" -----------------------------------------\n"
         << " [INFO] Starting pre-equilibration and density analysis workflow\n"
         << " Script    : " << argv[0] << "\n"
         << "===================================================================================================\n";

    try {
        modify_packmol(string(argv[2]) + ".inp",
                        target_mols, factor);

        modify_topology(string(top_env) + ".top",
                        target_mols, factor);
    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }

    cout << "✅ All files modified successfully!\n";
    return 0;
}
